### ISSUE TITLE
React state warning when joining a room

### ISSUE THREAD

**Issue by coreymaher** [twilio/twilio-video-app-react#497]
**Describe the bug**
When joining a video room the following react warning is displayed in the developer console:
```
Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in ToggleVideoButton (at DeviceSelectionScreen.tsx:114)
    in HiddenJs (created by WithWidth(HiddenJs))
    in WithWidth(HiddenJs) (created by Hidden)
    in Hidden (at DeviceSelectionScreen.tsx:112)
    in div (at DeviceSelectionScreen.tsx:111)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (at DeviceSelectionScreen.tsx:110)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (at DeviceSelectionScreen.tsx:109)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (at DeviceSelectionScreen.tsx:96)
```

**To Reproduce**
1. Start the app, fill out your name and room name
2. Click on 'Join Now' button
3. See the error in the developer console

**Expected behavior**
No react warnings or errors are displayed

**Screenshots**
N/A

**Environment (please complete the following information):**
 - OS: OSX 10.15.7
 - Browser: Firefox 87.0 and Chrome 89.0.4389.128
 - App Version: v0.3.1
 - SDK Version: 2.13.1
 - Node.js version: 12.20.1

**Additional context**
N/A

**Comment by timmydoza** [comment]
Thanks for the issue @coreymaher!

I can confirm that this issue exists. I dug into it a little bit, and it seems that components that use the `useDevices()` hook will trigger this warning if the component is mounted and unmounted quickly enough. Unfortunately, the promise returned by `navigator.mediaDevices.enumerateDevices()` (used by `useDevices`), isn't cancellable, so there's nothing that can be done to stop that async operation before a component that consumes `useDevices` is unmounted.

I'll look into fixing it someday, but after taking a look, it appears that this warning has no impact on end users. Additionally, the warning is suppressed when the app is build in production mode with `npm run build`, so most end users won't even see the warning if they happen to be looking at the console. I think this warning is just React's way of saying "hey, a component updated when it probably shouldn't. It might not be an issue, but take a look just in case".

I'll try to fix it, but since it has no impact on end users, this issue will be a low priority, so it may take some time before we can get to it. Thanks again for bringing it to our attention!

**Comment by coreymaher** [comment]
Thanks for looking into this, that sounds reasonable to me.

**Comment by hkumar0132** [comment]
@timmydoza   I am new here looking to make my first contribution to Twilio. This seems like a good start for me. Can you point me in the right direction? I'd love to work on this.

**Comment by ewanmellor** [comment]
The problem is not specifically that `navigator.mediaDevices.enumerateDevices()` is not cancellable, it is that the callback then goes on to set some React state (`setDeviceInfo` inside `useDevices` in this case).  You can avoid the warning as long as you don't do anything to your component state after the component has been unmounted.

I have a helper:

```
const useIsMounted = () => {
    const isMounted = useRef(false)
    useEffect(() => {
        isMounted.current = true
        return () => {
            isMounted.current = false
        }
    }, [])
    return isMounted
}
```

You can use it like this:

```
   const isMounted = useIsMounted()

    useEffect(() => {
        const getDevices = () =>
            getDeviceInfo(isAudioAllowed, isVideoAllowed).then(devices => {
                if (isMounted.current) {
                    setDeviceInfo(devices)
               }
            })

          ...
    })
```

(And if your build supports async, I'd prefer that over a chained `then` block.)

**Comment by timmydoza** [comment]
Hey everyone! I think I am going to close this issue. The message that we see in the console is just a warning, not an error. Sure, it can be an indication that there is a memory leak, but I don't think that's the case here (or if there is a leak, it is de minimis, as it only happens when connecting to a room, which is going to happen only once for most users). There is no impact on the end user, so putting the effort in to fix this won't improve the user experience for anyone. 

All that said, I think this is a very low-priority problem, and I don't think we'll have the bandwidth to fix it. We'll just need to [ignore the warning](https://i.redd.it/jnqaaz7ooim31.jpg) for now 🙂.

@ewanmellor - That's a really clever solution! But I don't think it's worth it to add it to the app. As the warning states, the React state update on the unmounted component is a no-op. So nothing happens when it is called. It's like calling a function like this:

```
const noop = () => {};
noop(); // Nothing happens!
```

Your solution adds additional code and complexity (which we would probably want to have unit tests for) to the application just to prevent a no-op from being executed. So in my mind it feels like an antipattern - it's doing something to prevent nothing from happening. I think it would be easier just to do nothing from the start 🙂.

Anyway, those are my thoughts. I'm happy to discuss this further, but I just wanted to give an update on the issue since it's getting a little stale. Thanks again for the issue @coreymaher! As always, please let me know if you have any other questions about anything.

**Comment by timmydoza** [comment]
It's interesting to note that this warning is removed in React 18. PR with a great explanation here: https://github.com/facebook/react/pull/22114
